File: go/array/67_two_sum.go
package array

// twoSum 两数之和 II - 输入有序数组
func twoSum(numbers []int, target int) []int {
	// 仅存在一个有效答案，非递减顺序排列 (不可以 重复使用相同的元素)
	// O(1) 的额外空间复杂度

	// 左右指针
	l, r := 0, len(numbers)-1

	for l < r {

		if numbers[l]+numbers[r] == target {
			return []int{l + 1, r + 1}
		}

		if numbers[l]+numbers[r] < target {
			l++
			r = len(numbers) - 1
		}

		if numbers[l]+numbers[r] > target {
			r--
		}
	}

	return []int{-1, -1}
}

File: rust/src/array/67_two_sum.rs


pub fn two_sum(numbers: Vec<i32>, target: i32) -> Vec<i32> {
    // 左右指针
    let (mut l, mut r) = (0, numbers.len() - 1);

    while l < r {
        if numbers[l] + numbers[r] == target {
            return vec![(l + 1) as i32, (r + 1) as i32];
        }

        // 小于target，在不存在此时以l 为加数的组合了，重置尝试下个组合
        if numbers[l] + numbers[r] < target {
            l += 1;
            r = numbers.len() - 1;
        }

        if numbers[l] + numbers[r] > target {
            r -= 1;
        }
    }
    vec![-1, -1]
}
